Loggly.class.nut

```diff
@@ -0,0 +1,138 @@
+class Loggly {
+    static version = [1,0,0];
+
+    static LOG_URL = "https://logs-01.loggly.com/bulk/%s/tag/%s/"
+
+    static LOG  = "LOG";
+    static WARN = "WARNING";
+    static ERR  = "ERROR";
+
+    _token = null;          // The customer token
+    _id = null;             // ID (default: agentID)
+    _tag = null;            // The log Tag (default: electricimp)
+    _timeout = null;        // Send frequency (default: 15s)
+    _debug = null;          // whether or not to server.log (default: true)
+
+    _url = null;            // The URL we send reqeusts to
+    _logString = null;      // The current set of logs that are queued for send
+
+    _onError = null;        // The onError handler for when there's a problem with the logs
+    _timer = null;          // The watchdog's timer object
+
+    constructor(token, options = {}) {
+        _token = token;
+
+        // Grab any settings
+        _id = "id" in options ? options.id : split(http.agenturl(), "/").pop();
+        _tag = "tag" in options ? options.tag : "electricimp"
+        _timeout = "timeout" in options ? options.timeout : 15;
+        _debug = "debug" in options ? options.debug : true;
+
+        // Generate the URL
+        _url = _generateUrl()
+
+        // initialize _logString
+        _logString = "";
+
+        // Start the send loop
+        _timer = imp.wakeup(_timeout, send.bindenv(this));
+    }
+
+    function log(msg, ...) {
+        _push(LOG, msg, vargv);
+    }
+
+    function warn(msg, ...) {
+        _push(WARN, msg, vargv);
+    }
+
+    function error(msg, ...) {
+        _push(ERR, msg, vargv);
+    }
+
+    function send() {
+        // reset the timer object
+        if (_timer != null) imp.cancelwakeup(_timer);
+        _timer = imp.wakeup(_timeout, send.bindenv(this));
+
+        // if there's nothing to log, we're done
+        if (_logString.len() == 0) return;
+
+        // Grab the logs, and clear
+        local logs = _logString;
+        _logString = "";
+
+        // Send the logs
+        http.post(_url, {}, logs).sendasync(function(resp) {
+            if (resp.statuscode != 200) {
+                // If an error occured, add the logs back in
+                _logString += logs;
+
+                if (_onError != null) {
+                    local __this = this;
+                    imp.wakeup(0, function() { __this._onError(resp); });
+                } else {
+                    server.error("Loggly send failed:");
+                    server.error("   " + resp.statuscode + " - " + resp.body);
+                }
+            } else {
+                server.log("success!");
+                // nothing
+            }
+        }.bindenv(this));
+    }
+
+    function len() {
+        return split(_logString, "\n").len();
```

Comment:
I was thinking we either needed to do this here, or when we added logs (because logging `"log1\nlog2"` logs two messages).. but since we don't allow devs to pass raw log messages to Loggly, we don't need to worry about that case.. 

I'll add a counter.